File: java/exercises/01-beginner/ex029_regex_validation/RegexValidationTest.java
package fewolearning.exercises.beginner.ex029_regex_validation;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

class RegexValidationTest {

    @Test
    void isValidEmailAcceptsAStandardShapedAddress() {
        assertTrue(RegexValidation.isValidEmail("ann.smith@example.com"));
    }

    @Test
    void isValidEmailRejectsAnAddressMissingAtSign() {
        assertFalse(RegexValidation.isValidEmail("ann.smith-example.com"));
    }

    @Test
    void isValidEmailRejectsAnAddressMissingADomainSuffix() {
        assertFalse(RegexValidation.isValidEmail("ann@example"));
    }

    @Test
    void isValidEmailRejectsBlank() {
        assertFalse(RegexValidation.isValidEmail(""));
    }

    @Test
    void extractDomainReturnsThePartAfterTheAtSign() {
        assertEquals("example.com", RegexValidation.extractDomain("ann.smith@example.com"));
    }

    @Test
    void extractDomainThrowsForAnInvalidEmail() {
        assertThrows(IllegalArgumentException.class, () -> RegexValidation.extractDomain("not-an-email"));
    }
}

File: kotlin/exercises/01-beginner/ex028_regex_validation/RegexValidationTest.kt
package fewolearning.exercises.beginner.ex028_regex_validation

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertNotNull

class RegexValidationTest {

    @Test
    fun isValidEmailIsTrueForAWellFormedAddress() {
        assertTrue(isValidEmail("john@example.com"))
    }

    @Test
    fun isValidEmailIsFalseForAMalformedAddress() {
        assertFalse(isValidEmail("not-an-email"))
        assertFalse(isValidEmail("missing@domain"))
    }

    @Test
    fun userAndDomainDestructuresAValidAddress() {
        val result = userAndDomain("john@example.com")

        assertNotNull(result)
        val (user, domain) = result!!
        assertEquals("john", user)
        assertEquals("example.com", domain)
    }

    @Test
    fun userAndDomainReturnsNullForAnInvalidAddress() {
        assertNull(userAndDomain("not-an-email"))
    }
}
